### ISSUE TITLE
bundler installs new gems when it shouldn't

### ISSUE THREAD

**Issue by axlekb** [bundler/bundler#3541]
I'm having an issue with `bundle install` on `1.9.2` ignoring `Gemfile.lock` gem versions for secondary sources.

##### Simplified setup (not actually tested, but as an example)
```ruby
source 'https://rubygems.org'
gem 'bugsnag'
gem 'rails-assets-datatables', source: 'https://rails-assets.org'
```

If a more recent version of the gem is available, when I `bundle install` with no version set for `rails-assets-datatables`, it is updated to the latest version in Gemfile.lock.

Additionally, with no version in the Gemfile, all of the secondary source gems are updated on a `bundle update`.

**Comment by indirect** [comment]
Sorry, I don't think I understand the problem. Can you give an example that includes the versions available, the versions that you expect, and the versions you actually get?

Also, can you provide example code that shows the problem? That makes it about 100x more likely that we can fix it.

**Comment by axlekb** [comment]
Alright... here we go.

##### Initial Setup (each gem is one version behind latest available)
```ruby
### Gemfile
source 'https://rubygems.org'

gem 'mime-types', '2.4.2'
source 'https://rails-assets.org' do
  gem 'rails-assets-numeraljs', '1.5.2'
end
```
##### Initial Gemfile.lock
```ruby
### Gemfile.lock
GEM
  remote: https://rubygems.org/
  remote: https://rails-assets.org/
  specs:
    mime-types (2.4.2)
    rails-assets-numeraljs (1.5.2)

PLATFORMS
  ruby

DEPENDENCIES
  mime-types (= 2.4.2)
  rails-assets-numeraljs (= 1.5.2)!
```

##### Remove version restrictions, run  `bundle update mime-types`
```
# Gemfile
source 'https://rubygems.org'

gem 'mime-types'
source 'https://rails-assets.org' do
  gem 'rails-assets-numeraljs'
end
```
##### Resulting Gemfile.lock showing `rails-assets-numeraljs` updated to `1.5.3`
```ruby
# Gemfile.lock
GEM
  remote: https://rubygems.org/
  remote: https://rails-assets.org/
  specs:
    mime-types (2.4.3)
    rails-assets-numeraljs (1.5.3)   # <---- WHY DID THIS GET UPDATED?

PLATFORMS
  ruby

DEPENDENCIES
  mime-types
  rails-assets-numeraljs!
```

**Comment by indirect** [comment]
So... while I can reproduce this on Bundler 1.9.1, I cannot reproduce this on 1.9.2:

```
$ bundle update mime-types
Fetching gem metadata from https://rails-assets.org/..
Fetching version metadata from https://rails-assets.org/.
Fetching gem metadata from https://rails-assets.org/..
Fetching gem metadata from https://rubygems.org/.........
Fetching version metadata from https://rails-assets.org/..
Fetching version metadata from https://rubygems.org/..
Resolving dependencies...
Using rails-assets-numeraljs 1.5.2
Using bundler 1.9.2
Installing mime-types 2.4.3 (was 2.4.2)
Bundle updated!
```

Are you _completely_ sure you're using 1.9.2?

**Comment by axlekb** [comment]
Positive that I have `1.9.2`, just uninstalled and reinstalled.  Can still reproduce locally.  Seems that it actually has to do with with having the later gem version available locally (which happened automatically with the bug < `1.9.2`).  Since the `Gemfile.lock` does not keep version numbers of the alternate sources, it just installs the maximum version.  So if `numeraljs 1.5.3` is available, it gets installed.  Is there a reason why the Gemfile.lock doesn't keep version numbers for alternate source gems?

**Comment by axlekb** [comment]
Can this be reopened?

**Comment by indirect** [comment]
I tried running your reproduction case, and it seems to be fixed in 1.9.2. Can you provide a new reproduction case that demonstrates the issue?
